unstable: Implement HostListener mechanism for CPP-RS Driver

wprzytula opened the pull request:
## What's done

### Introduced `unstable-host-listener` feature

The feature gates the whole `host_listener.rs` module. However, even with feature disabled, driver now emits logs about nodes changing their connectivity status, as well as joining/leaving the cluster.

### Added `HostListener` API

`HostListener` is a trait which allow subscribing to node-related (topology and status) events. It involves one method, `on_event`, which takes the context (for now: host ID and SocketAddr, easily extensible in the future due to lack of public fields and getter-based API) and the event kind (ADD, REMOVE, UP, or DOWN).

### Implemented connectivity change notification mechanism in `PoolRefiller`

`PoolRefiller` now notifies `ClusterWorker` through a channel that an endpoint is again reachable or no longer reachable. This is done by sending a pair `(<host_id>, <event_kind>)` (with event represented by `ConnectivityChangeEvent` enum) into the channel.

### Added node status tracking map to `Cluster`, based on connectivity

Now `ClusterWorker` stores a connectivity status marker for every node in the cluster, which is updated by `ClusterWorker` each time the connection pool that corresponds to the given node notifies it about the connectivity change.

### `ClusterWorker` notifies `HostListener` each time status marker in the tracker map changes

This is a notification deduplication mechanism: should any `PoolRefiller` somehow send two same notifications in a row, only the first will result in `HostListener::on_event` callback fired.

### `ClusterWorker` notifies `HostListener` about every node added/removed from the cluster metadata

Now, upon each metadata refresh:
1. every node that was present in old `ClusterState` but is absent from new `ClusterState` is reported as REMOVED.
2. every node that was absent from old `ClusterState` but is present in new `ClusterState` is reported as ADDED.

Additionally:
1. nodes are reported as DOWN before REMOVED (if hadn't been reported as DOWN earlier).
2. nodes are reported as UP after ADDED.
    - If the pool fails to open any connections to the new node, it will later be marked as DOWN. This mimics CPP Driver's behaviour.
3. nodes filtered out by `HostFilter` have no events reported at all. This mimics CPP Driver's behaviour.


## What has been tested

### Manual testing

I built CPP-RS Driver on top of this branch and wrote the bridge to make `host_listener.c` example run. Then, I ran both CPP Driver and CPP-RS Driver in parallel and played with the cluster: stopped and started nodes, added and removed nodes. Broke control connection or not in all of these scenarios. In all situations the Rust implementation behave the same _or better_ than the CPP Driver, which serves as a reference.

### Automated CCM test

I wrote an integration test using CCM bridge (thank you @dkropachev and @Lorak-mmk for making this awesome testing technique possible!). It includes a test cases for quite convoluted cluster dynamics (nodes stopped and started, control connection broken and reestablished on another node, nodes removed and added), which makes me quite confident about correctness of the implementation.


## TODO

### More Test Cases

There likely are corner cases that we haven't been testing yet. Ideas include:
- node changing its IP
- ...


Fixes: #1480

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

Lorak-mmk commented on scylla/src/policies/host_listener.rs:
```diff
@@ -0,0 +1,81 @@
+    /// 2. When a CQL event notifies about a new node leaving the cluster.
+    /// 3. When a CQL event notifies about a known node being added. Then this event is fired
+    ///    right before the `host_added` event.
+    Removed,
```
Huh, so this comment now tells me that events really are taken into consideration. Interesting, I definitely need to read the implementation.

wprzytula replied:
This is as well outdated.

Lorak-mmk commented on scylla/src/policies/mod.rs:
```diff
@@ -17,6 +17,8 @@
 pub mod address_translator;
 pub mod host_filter;
+// #[cfg(cpp_rust_unstable)]
+pub mod host_listener;
```
Why is this commented out?

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -268,6 +268,10 @@ pub struct SessionConfig {
 
+    // #[cfg(cpp_rust_unstable)]
+    /// Optional listener for host events (ADD, REMOVE, UP, DOWN).
+    pub host_listener: Option<Arc<dyn crate::policies::host_listener::HostListener>>,
```
ditto

Lorak-mmk commented on scylla/src/network/connection_pool.rs:
```diff
@@ -1187,6 +1187,16 @@ struct OpenedConnectionEvent {
+    /// The last working connection to the node was lost.
+    Lost { address: NodeAddr },
+}
+
```
Commit: " connection_pool: introduce ConnectivityChangeEvent "

It is surprising that NodeAddr is used for identification, given all the issues and ambiguities we have with identifying nodes this way. Maybe the reason for that will become clear in later commits, idk.

wprzytula replied:
I use it because this is an internal API, so I preferred to keep the additional information that NodeAddr has compared to SocketAddr. It's up to the upper layer to convert it to SocketAddr, which it btw does.

wprzytula replied:
Shall we consider this resolved?

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -902,6 +902,7 @@ impl Session {
             can_use_shard_aware_port: !config.disallow_shard_aware_port,
+            connectivity_events_sender: None, // This will be set by `Cluster::new`.
         };
 
```
This becomes yet another way signal out of connection pool - we already had `pool_updated_notify` and `pool_empty_notifier`. Instead of introducing third one, can we do some unification?

`pool_empty_notifier` sounds like something really close to what we need - its just that we need to send both when it becomes empty, and starts being empty. Current users of `pool_empty_notifier` could ignore the "stops being empty" case.

wprzytula replied:
As discussed, possible unification of these notification channels is out of scope of this PR.

Lorak-mmk commented on scylla/src/network/connection_pool.rs:
```diff
@@ -816,6 +821,23 @@ impl PoolRefiller {
+                    // Ignore failure to send. If there are no listeners, it's fine.
+                    let event = ConnectivityChangeEvent::Established { address: endpoint };
+                    let _ = connectivity_notifier.send(event);
+                }
```
I wonder, why are you performing the notification here?
I see that below there are some checks which could eliminate the connection from being added.
In the case it is accepted, `update_shared_conns` will be called, which also has notification logic added by you. Why duplicate it?

wprzytula replied:
`update_shared_conns` only has logic for DOWN, not for UP.

wprzytula commented on scylla/src/policies/host_listener.rs:
```diff
@@ -0,0 +1,81 @@
+    /// This happens in the following cases:
+    /// 1. When the control connection discovers a new peer.
+    /// 2. When a CQL event notifies about a new node joining the cluster.
+    Added,
```
It should. I forgot to update the docstring.

wprzytula commented on scylla/src/policies/host_listener.rs:
```diff
@@ -0,0 +1,81 @@
+    pub(crate) host_id: Uuid,
+    pub(crate) addr: SocketAddr,
+}
+
```
OK, will do. I first added it, but then removed due to `pub(crate)` visibility.

Lorak-mmk commented on scylla/src/network/connection_pool.rs:
```diff
@@ -816,6 +821,23 @@ impl PoolRefiller {
+                    let event = ConnectivityChangeEvent::Established { address: endpoint };
+                    let _ = connectivity_notifier.send(event);
+                }
+
```
Ok

wprzytula commented on scylla/src/cluster/worker.rs:
```diff
@@ -456,6 +464,131 @@ impl ClusterWorker {
+                self.0.address.into_inner().hash(state);
+            }
+        }
+
```
I fear that it will be dysfunctional. I can't even properly reason about such scenarios. For now, I'm relying on the fact that there's no address translation in CPP-RS driver.

wprzytula commented:
v1.2:
- improved the integration test.
- fixed bug in Node: `inherit_with_ip_changed` must copy the last UP/DOWN marker state. This is irrelevant, though, because I'm planning to remove the marker from Node as discussed with @Lorak-mmk in person.

wprzytula commented:
v1.4:
- removed UP/DOWN deduplication marker from Node; instead, ClusterWorker now owns a node status map for this purpose.
- moved ADDED/REMOVED condition handling to `ClusterData::new` (although made it call ClusterWorker's code back for better responsibility separation).
- fixed handling of two corner cases:
  - when a node changes its IP,
  - when one node is removed and another is added with the same IP.

Lorak-mmk commented on scylla/src/cluster/worker.rs:
```diff
@@ -46,11 +47,27 @@ impl std::fmt::Debug for ClusterNeatDebug<'_> {
+    /// Maps host_id to connectivity status.
+    /// Used to track node status in order to deduplicate HostListener events.
+    node_status: HashMap<Uuid, (NodeAddr, NodeConnectivityStatus)>,
+
```
ConnectivityChangeEvents no longer identify nodes by Uuid. Is there any reason now to have addr in this map?

wprzytula replied:
Good observation. Will remove this.

Lorak-mmk reviewed (changes requested):
The tests look great!

Lorak-mmk commented on scylla/src/cluster/worker.rs:
```diff
@@ -467,6 +502,218 @@ impl ClusterWorker {
+            &dyn HostListener,
+        >,
+        node_status: &mut HashMap<Uuid, NodeConnectivityStatus>,
+    ) {
```
I'm not a fan of those `#[cfg(...)]` now being everywhere now. Idea: we can make HostListener exist unconditionally, but without the feature it will not be present in SessionConfig, and it will not be pub.
To make it non-pub, you can make the module containing it private, the items themselves pub, and add a pub re-export guarded by cfg.

Lorak-mmk commented on scylla/tests/integration/utils.rs:
```diff
@@ -450,7 +450,13 @@ pub(crate) async fn execute_unprepared_statement_everywhere(
+        all(feature = "openssl-010", feature = "rustls-023")
+    )),
+    expect(dead_code)
+)]
```
When I see such declarations, I wonder if we can just `allow(dead_code)` for whole `ccm` module, and utils like this one. Probably not a good idea, idk.

wprzytula replied:
The problem of such declarations will disappear once we start using CCM for testing stable core features of the driver. Let's do it :)

wprzytula commented:
> Uh, I thought it was going to be an approve, but I found one more thing, sorry :(

Done.
